posthog/temporal/batch_exports/destination_tests.py

```diff
@@ -0,0 +1,466 @@
+    to check whether we can create a table, it makes no sense to run the second
+    test step if the first one fails. Future revisions of this interface could
+    expand the hierarchy to allow for multiple paths (a tree), but for now a simple
+    list is sufficient.
+
+    Attributes:
+        steps: A property returning a sequence of steps to run.
+    """
+
+    @abc.abstractmethod
+    def configure(self, **kwargs):
+        """Method to configure a concrete test.
+
+        By "configure" I mean setting any attributes required to initialize and/or
+        run test steps.
+
+        Subclasses should override this to set any attributes. This decoupling of
+        configuration from initialization allows us to serialize a `DestinationTest`
+        without needing to configure it.
+        """
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def steps(self) -> collections.abc.Sequence[DestinationTestStep]:
+        """Property returning a sequence of steps to run.
+
+        Subclasses should override this with their required test steps.
+        """
+        raise NotImplementedError
+
+    def run_step(self, step: int) -> DestinationTestStep:
+        """Run the test step at index `step`."""
+        test_step = self.steps[step]
+        step_result = async_to_sync(test_step.run)()
+
+        test_step.result = step_result
+        return test_step
+
+    def as_dict(self) -> dict[str, list[DestinationTestStepDict]]:
+        """Serialize this as a dictionary."""
+        return {"steps": [step.as_dict() for step in self.steps]}
+
+
+class S3CheckBucketExistsTestStep(DestinationTestStep):
+    """Test whether an S3 bucket exists and we can access it.
+
+    This test could not be broken into two as the bucket not existing and not having
+    permissions to access it looks the same from our perspective.
+
+    Attributes:
+        bucket_name: The bucket we are checking.
+        region: Region where the bucket is supposed to be.
+        endpoint_url: Set for S3-compatible destinations.
+        aws_access_key_id: Access key ID for the bucket.
+        aws_secret_access_key: Secret access key for the bucket.
+    """
+
+    def __init__(
+        self,
+        bucket_name: str | None = None,
+        region: str | None = None,
+        endpoint_url: str | None = None,
+        aws_access_key_id: str | None = None,
+        aws_secret_access_key: str | None = None,
+    ) -> None:
+        super().__init__(
+            name="Check S3 bucket exists", description="Verify the configured S3 bucket exists and we can access it"
+        )
+        self.bucket_name = bucket_name
+        self.region = region
+        self.endpoint_url = endpoint_url
+        self.aws_access_key_id = aws_access_key_id
+        self.aws_secret_access_key = aws_secret_access_key
+
+    async def run(self) -> DestinationTestStepResult:
+        """Run this test step."""
+        import aioboto3
```

Comment:
Same question about imports here

Reply:
In this particular case, I liked the fact that the imports are included as part of the test itself.

Reply:
for organizational reasons?

Reply:
In case we ever introduce errors that are raised on import, I rather that cause a test failure just for the relevant test.

Reply:
I guess that's a valid point, although ideally we catch these in our unit tests